openedx/features/course_experience/views/welcome_message.py

```diff
@@ -0,0 +1,49 @@
+"""
+View logic for handling course welcome messages.
+"""
+
+from django.template.loader import render_to_string
+
+from courseware.courses import get_course_info_section_module, get_course_with_access
+from opaque_keys.edx.keys import CourseKey
+from openedx.core.djangoapps.plugin_api.views import EdxFragmentView
+from web_fragments.fragment import Fragment
+
+
+class WelcomeMessageFragmentView(EdxFragmentView):
+    """
+    A fragment that displays a course's welcome message.
+    """
+    def render_to_fragment(self, request, course_id=None, **kwargs):
+        """
+        Renders the welcome message fragment for the specified course.
+
+        Returns: A fragment, or None if there is no welcome message.
+        """
+        course_key = CourseKey.from_string(course_id)
+        course = get_course_with_access(request.user, 'load', course_key, check_if_enrolled=True)
+        welcome_message_html = self.welcome_message_html(request, course)
+        if not welcome_message_html:
+            return None
+
+        context = {
+            'welcome_message_html': welcome_message_html,
+        }
+
+        html = render_to_string('course_experience/welcome-message-fragment.html', context)
+        return Fragment(html)
+
+    @classmethod
+    def welcome_message_html(cls, request, course):
+        """
+        Returns the course's welcome message or None if it doesn't have one.
+        """
+        # TODO: is there a more efficient way to get the first course update?
```

Comment:
Yes, this is what lead to the extra queries (interestingly some change I made reduced it to an extra 2 rather than 5). We are already batching, so I don't know that there's much else we can do, but maybe we can find out when we profile the code.

For now I'll remove the TODO.